components/camel-mail/src/test/java/org/apache/camel/dataformat/mime/multipart/MimeMultipartDataFormatTest.java

```diff
@@ -351,6 +351,41 @@ public void unmarshalInlineHeadersNonMimeBodyStream() throws UnsupportedEncoding
         assertEquals("This is not a MIME-Multipart", bodyStr);
     }
 
+    @Test
+    public void attachmentReadOnce() throws IOException {
+        String attContentType = "text/plain";
+        String attText = "Attachment Text";
+        InputStream attInputStream = new ByteArrayInputStream(attText.getBytes());
+        String attFileName = "Attachment File Name";
+        in.setBody("Body text");
+        in.setHeader(Exchange.CONTENT_TYPE, "text/plain;charset=iso8859-1;other-parameter=true");
+        in.setHeader(Exchange.CONTENT_ENCODING, "UTF8");
+        Map<String, String> headers = new HashMap<String, String>();
+        headers.put("Content-Description", "Sample Attachment Data");
+        headers.put("X-AdditionalData", "additional data");
+        CountingByteArrayDataSource attachmentDs = new CountingByteArrayDataSource(attInputStream, attContentType);
+        addAttachment(attachmentDs, attFileName, headers);
+        Exchange result = template.send("direct:roundtrip", exchange);
+        Message out = result.getOut();
+        assertEquals("Body text", out.getBody(String.class));
+        assertThat(out.getHeader(Exchange.CONTENT_TYPE, String.class), startsWith("text/plain"));
+        assertEquals("UTF8", out.getHeader(Exchange.CONTENT_ENCODING));
+        assertTrue(out.hasAttachments());
+        assertEquals(1, out.getAttachmentNames().size());
+        assertThat(out.getAttachmentNames(), hasItem(attFileName));
+        Attachment att = out.getAttachmentObject(attFileName);
+        DataHandler dh = att.getDataHandler();
+        assertNotNull(dh);
+        assertEquals(attContentType, dh.getContentType());
+        InputStream is = dh.getInputStream();
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        IOHelper.copyAndCloseInput(is, os);
+        assertEquals(attText, new String(os.toByteArray()));
+        assertEquals("Sample Attachment Data", att.getHeader("content-description"));
+        assertEquals("additional data", att.getHeader("X-AdditionalData"));
+        assertEquals(1, attachmentDs.readCounts); // Fails - input is read twice
```

Comment:
Thanks for posting a PR @tdudgeon . As I get it, the test is not intended to produce any exception but rather expect that `CountingByteArrayDataSource.getInputStream(...)` be called once. Is it really an issue if it is called twice ?

Reply:
@tdudgeon , do you have any input to @aldettinger 's question?

Reply:
@aldettinger Yes, it can matter (and sorry for slow response). Some types of DataSource cannot be read repeatedly so you get an exception when it read a second time. 
I'm sure this was discussed on the users mailing list, but I can't find the topic right now :-(

Reply:
I think this is http://camel.465427.n5.nabble.com/MIME-Mutipart-DataFormat-streaming-td5789219.html

Reply:
@tdudgeon , @aldettinger any update on this topic?

Reply:
@onderson Not sure what update you want on this. The code does seem to read the InputStream twice, and that can cause problems when the stream can't be read multiple times.

Reply:
By just looking into your failing test case, it is hard to estimate your use case from my side. And as of an initial look, it seems like an issue/limitation about ByteArrayDataSource and maybe mail component needs another implementation of DataSource so user will be able to read attachments multiple times? Is this the solution you need and which requires such exploration on the component and mail API as well?

Reply:
@tdudgeon, IMHO the proposed test does not yet exhibit an issue as a valid DataSource.getInputStream() implementation would return [a new stream positioned at the beginning of the data each time it is called](https://docs.oracle.com/javase/8/docs/api/javax/activation/DataSource.html#getInputStream--).

There I have 2 questions:
Could it be that you faced an issue involving a wrong DataSource implementation ?
Or, was your issue linked to a DataSource returning an input stream that don't support mark / reset ?